_posts/2016-05-11-dry-your-test-unit.markdown

```diff
@@ -0,0 +1,149 @@
+---
+layout: post
+title: "DRY your tests"
+date: 2016-05-13 09:01:00
+categories: ["rails", "dry"]
+author: "etagwerker"
+---
+
+I'm big fan of having small classes. I'm not a big fan of having huge specs for
+a small class/object. Every time I see an opportunity to
+[DRY](http://c2.com/cgi/wiki?DontRepeatYourself) my specs, I take it.
+
+Today I wrote a spec to make sure that we gracefully ignore SPAMmy
+contact requests in the Ombu Labs
+[contact page](http://www.ombulabs.com/contact). It initially looked like this:
+
+```ruby
+test "gracefully ignores spammy requests with valid attributes" do
+  @valid_contact = contacts(:two)
+  attributes = @valid_contact.attributes
+  attributes.merge!(email_confirmation: @valid_contact.email)
+
+  assert_no_difference("Contact.count") do
+    post :create, contact: attributes, format: 'js'
+  end
+
+  assert_response :success
+end
+```
+
+The new behavior adds a simple SPAM trap field that bots will usually fall for.
+If a bot is submitting the `email_confirmation` field, then it is SPAM and it
+gracefully ignores the request.
+
+The test only tests the scenario where the bot is performing an AJAX request.
+Then I thought that SPAM bots might try to submit a non-AJAX `html` request.
+So I wrote some more:
+
+```ruby
+test "gracefully ignores spammy requests with valid attributes (AJAX)" do
+  @valid_contact = contacts(:two)
+  attributes = @valid_contact.attributes
+  attributes.merge!(email_confirmation: @valid_contact.email)
+
+  assert_no_difference("Contact.count") do
+    post :create, contact: attributes, format: 'js'
+  end
+
+  assert_response :success
+end
+
+test "gracefully ignores spammy requests with valid attributes (HTML)" do
+  @valid_contact = contacts(:two)
+  attributes = @valid_contact.attributes
+  attributes.merge!(email_confirmation: @valid_contact.email)
+
+  assert_no_difference("Contact.count") do
+    post :create, contact: attributes, format: 'html'
+  end
+
+  assert_response :success
+end
+```
+
+Now that is **ridiculous**, why should I **copy/paste** this? There is only one
+parameter (`format`) that varies between them.
+
+So, I [refactored](http://c2.com/cgi/wiki?RefactorMercilessly) the code to
+look like this:
+
+```ruby
+["js", "html"].each do |format|
+  test "gracefully ignores spammy requests with valid attributes" do
+    @valid_contact = contacts(:two)
+    attributes = @valid_contact.attributes
+    attributes.merge!(email_confirmation: @valid_contact.email)
+
+    assert_no_difference("Contact.count") do
+      post :create, contact: attributes, format: format
+    end
+
+    assert_response :success
+  end
+end
+```
+
+The code above doesn't *really* work. It raises this exception:
+
+    /Users/etagwerker/.rvm/gems/ruby-2.2.1@ombulabs-com/gems/activesupport-3.2.17/lib/active_support/testing/declarative.rb:28:in `test': test_should_gracefully_ignore_spammy_requests_with_valid_attributes is already defined in ContactsControllerTest (RuntimeError)
```

Comment:
@etagwerker should we maybe wrap this exception with backticks or a pre tag or something?